### PR TITLE
Avoid interrupt inflight requests after a new socket connect failed

### PR DESCRIPTION
**Pull request by domodwyer**
See https://github.com/go-mgo/mgo/pull/437

### CODE REVIEW

**Comment by domodwyer** [comment]
Thanks @carter2000, much appreciated.

File: server.go [hunk 1]
```diff
@@ -187,6 +187,16 @@ func (server *mongoServer) Connect(timeout time.Duration) (*mongoSocket, error)
...
+	server.close(false)
+}
+
+// CloseIdle closing all sockets that are idle,
+// sockets currently in use will be closed after idle.
+func (server *mongoServer) CloseIdle() {
+	server.close(true)
+}
+
+func (server *mongoServer) close(idle bool) {
```
Comment (tadukurow) [inline server.go]: func (server *mongoServer) close(waitForIdle bool)

Comment (carter2000) [inline server.go]: thanks, I have changed it as you suggest.

File: socket.go [hunk 2]
```diff
@@ -316,6 +320,19 @@ func (socket *mongoSocket) Close() {
 	socket.kill(errors.New("Closed explicitly"), false)
 }
```
Comment (vgalu) [inline socket.go]: godoc?

**Comment by domodwyer** [comment]
Thanks @carter2000 - we'll be testing a build of this and then we'll merge it into master.

Cheers!

**Pull request merged**